react/getting_started_with_react/keys_in_react.md

```diff
@@ -0,0 +1,126 @@
+2. we hunt down that specific items that were changed and only re-render those.
+
+Assuming we want to hunt down that one specific item that was changed and NOT re-render the entire list. We need something to track that specific item. We can track down a specific item by using a `key`.
+
+When the list is updated for whatever reason, (either from a server or a user interaction), React matches the `keys` of each of the previous list to the updated list. If there were any changes, React will only update the items that have changed.
+
+As long as `keys` remain consistent and unique, React can handle the DOM effectively and efficiently.
+
+### Using keys
+
+Keys are passed into the component or a DOM element as a prop. You should already be familiar with the syntax.
+
+~~~jsx  
+<Component key={keyValue} />
+//or
+<div key={keyValue} />
+~~~
+
+<span id="keys-from-data">Now that we know the syntax, the next question is: what should be used as a key? Ideally they should be some identifier that is unique to each item in the list. Most databases assign a unique id to each entry, so you shouldn't have to worry about assigning and id yourself. If you are defining data yourself, it is good practice to assign a unique `id` to each item. You may use the [uuid](https://www.npmjs.com/package/uuid) package to generate a unique id. Let's look at an example:</span>
+
+~~~jsx
+// a list of todos, each todo object has a task and an id
+const todos = [
+  { task: "mow the yard", id: uuid() }, 
+  { task: "Work on Odin Projects", id: uuid() },
+  { task: "feed the cat", id: uuid() },
+];
+
+function TodoList() {
+  return (
+    <ul>
+      {todos.map((todo) => (
+        // here we are using the already generated id as the key.
+        <li key={todo.id}>{todo.task}</li>
+      ))}
+    </ul>
+  ) 
+}
+~~~
+
+<span id="index-as-key">Additionally, if the list is static and will remain unchanged throughout the application's life, we can use the array index as a key. Though you should be careful when working with `index` as key.</span>
+
+~~~jsx
+const months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'];
+
+function MonthList(){
+  return (
+    <ul>
+      {/* here we are using index as key */}
+      {months.map((month, index) => (<li key={index}>{month}</li>))}
+    </ul>
+  )
+}
+~~~
+
+<span id="anti-pattern">Keys are straightforward to use, though there is an anti-pattern you should be aware of. Keys should never be generated on the fly. Using `key={Math.random()}` or `key={uuid()}` _while_ rendering the list defeats the purpose of the list, as now a new `key` will ge created for every render of the list. As shown in the above example, `key` should be inferred from the data itself.</span>
+
+~~~jsx
+const todos = [
+  { task: "mow the yard", id: uuid() }, 
+  { task: "Work on Odin Projects", id: uuid() },
+  { task: "feed the cat", id: uuid() },
```

Comment:
Don't really need the `id` property in these objects for the purpose of this example